lib/tree.rb

```diff
@@ -16,40 +16,162 @@ def initialize
+      current.left = add_helper(current.left, new_node, key)
+    else
+      current.right = add_helper(current.right, new_node, key)
+    end
+    return current
+  end
+
+
+  # Time Complexity: log(n)
+  # Space Complexity: log(n)
   def find(key)
-    raise NotImplementedError
+    if @root.nil?
+      return nil
+    else
+      find_helper(key, @root)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(key, current)
+    return nil if current.nil?
+    if current.key == key
+      return current.value
+    elsif key > current.key
+      find_helper(key, current.right)
+    else
+      find_helper(key, current.left)
+    end
+  end
+
+  def find_node(key)
+    if @root.nil?
+      return nil
+    else
+      find_node_helper(key, @root)
+    end
+  end
+
+  def find_node_helper(key, current)
+    return nil if current.nil?
+    if current.key == key
+      return current
+    elsif key > current.key
+      find_node_helper(key, current.right)
+    else
+      find_node_helper(key, current.left)
+    end
+  end
+
+  def delete(key)
+    node = find_node(key)
+    unless node.nil?
+      remove(node)
+    end
+  end
+
+  def remove(node)
+    if node.left.nil? && node.right.nil?
+      node = nil
+    elsif !node.left.nil? && node.right.nil?
+      node = node.left
+    elsif node.left.nil? && !node.right.nil?
+      node = node.right
+      puts node.value
+    else
+      min_node = find_min_node(node.right)
+      node = min_node
+      min_node = nil
+    end
+  end
+
+  def find_min_node(node)
+    if node.left.nil?
+      min_node = node
+      return min_node
+    else
+      find_min_node(node.left)
+    end
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    return inorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def inorder_helper(current, list)
+    return list if current.nil?
+
+    inorder_helper(current.left, list)
+    list << { key: current.key, value: current.value}
+    inorder_helper(current.right, list)
+    
+    return list
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    return preorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, list)
+    return list if node.nil?
+
+    list << { key: node.key, value: node.value}
+    preorder_helper(node.left, list)
+    preorder_helper(node.right, list)
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    return postorder_helper(@root, [])
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(node, list)
+    return list if node.nil?
+
+    postorder_helper(node.left, list)
+    postorder_helper(node.right, list)
+    list << { key: node.key, value: node.value}
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(log n)
   def height
```

Comment:
This method isn't actually working.  See my note below. 

Think about this:

If the node is `nil` then return 0
If the node is not 0, find the height of the left and the right.

Return the height of the bigger subtree and add one (for the current node).